provider/cmd/crd2pulumi/schema.go

```diff
@@ -0,0 +1,303 @@
+}
+
+// GetObjectTypeSpecs generates types for each versioned schema. Returns a
+// mapping from each type's name in the format of
+// "{.spec.group}/{.spec.versions[*].name}:{.spec.names.kind}" to its proper
+// pschema.ObjectTypeSpec.
+func GetObjectTypeSpecs(versions map[string]map[string]interface{}, name, kind string) map[string]pschema.ObjectTypeSpec {
+	objectTypeSpecs := map[string]pschema.ObjectTypeSpec{}
+	for version, schema := range versions {
+		baseRef := fmt.Sprintf("crds:%s/%s:%s", name, version, kind)
+		addType(schema, baseRef, objectTypeSpecs)
+		// Adds "Args" to the baseRef name
+		newBaseRef := baseRef + "Args"
+		objectTypeSpecs[newBaseRef] = objectTypeSpecs[baseRef]
+		objectTypeSpecs[newBaseRef].Properties["metadata"] = pschema.PropertySpec{
+			TypeSpec: pschema.TypeSpec{
+				Ref: ObjectMetaRef,
+			},
+		}
+		delete(objectTypeSpecs, baseRef)
+	}
+
+	return objectTypeSpecs
+}
+
+// addType converts the given OpenAPI `schema` to a ObjectTypeSpec and adds it
+// to the `types` map under the given `name`. Recursively converts and adds all
+// nested schemas as well.
+func addType(schema map[string]interface{}, name string, types map[string]pschema.ObjectTypeSpec) {
+	properties, foundProperties, _ := unstruct.NestedMap(schema, "properties")
+	if !foundProperties {
+		return
+	}
+
+	description, _, _ := unstruct.NestedString(schema, "description")
+	schemaType, _, _ := unstruct.NestedString(schema, "type")
+	required, _, _ := unstruct.NestedStringSlice(schema, "required")
+
+	propertySpecs := map[string]pschema.PropertySpec{}
+	for propertyName := range properties {
+		propertySchema, _, _ := unstruct.NestedMap(properties, propertyName)
+		propertyDescription, _, _ := unstruct.NestedString(propertySchema, "description")
+		defaultValue, _, _ := unstruct.NestedFieldNoCopy(propertySchema, "default")
+		propertySpecs[propertyName] = pschema.PropertySpec{
+			TypeSpec:    getTypeSpec(propertySchema, name+strings.Title(propertyName), types),
+			Description: propertyDescription,
+			Default:     defaultValue,
+		}
+	}
+
+	types[name] = pschema.ObjectTypeSpec{
+		Type:        schemaType,
+		Properties:  propertySpecs,
+		Required:    required,
+		Description: description,
+	}
+}
+
+// Replaces the hyphens in the x-kubernetes-... fields with underscores
+func underscoreFields(schema map[string]interface{}) {
+	for field, val := range schema {
+		for _, hyphenedField := range HyphenedFields {
+			if field == hyphenedField {
+				delete(schema, field)
+				underScoredField := strings.ReplaceAll(field, "-", "_")
+				schema[underScoredField] = val
+			}
+		}
+		subSchema, ok := val.(map[string]interface{})
+		if ok {
+			underscoreFields(subSchema)
+		} else {
+			subSchemaSlice, ok := val.([]interface{})
+			if ok {
+				for _, genericSubSchema := range subSchemaSlice {
+					subSchema, ok = genericSubSchema.(map[string]interface{})
+					if ok {
+						underscoreFields(subSchema)
+					} else {
+						break
+					}
```

Comment:
I don't think you want to `break` here, otherwise, what if there are other values in the slice that need to be underscored?

Also, this function could be cleaned up a bit. Something like (untested):

```go
func underscoreFields(schema map[string]interface{}) {
	for field, val := range schema {
		if _, has := HyphenedFields[field]; has {
			delete(schema, field)
			underScoredField := strings.ReplaceAll(field, "-", "_")
			schema[underScoredField] = val
		}
		if subSchema, ok := val.(map[string]interface{}); ok {
			underscoreFields(subSchema)
		} else if subSchemaSlice, ok := val.([]interface{}); ok {
			for _, genericSubSchema := range subSchemaSlice {
				if subSchema, ok = genericSubSchema.(map[string]interface{}); ok {
					underscoreFields(subSchema)
				}
			}
		}
	}
}
```

Reply:
I think my assumption with the OpenAPI schema was that map[string]interface{}s had to fill up the entire slice, so if we found a single interface{} that couldn't be casted to map[string]interface{}, then we could early exit. However I'm not 100% sure on this, so I'll drop the `break` statement.